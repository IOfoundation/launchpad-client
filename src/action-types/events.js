const EventsTypes = {
  GET_ALL_EVENTS_REQUEST_START: 'GET_ALL_EVENTS_REQUEST_START',
  GET_ALL_EVENTS_REQUEST_SUCCESS: 'GET_ALL_EVENTS_REQUEST_SUCCESS',
  GET_ALL_EVENTS_REQUEST_ERROR: 'GET_ALL_EVENTS_REQUEST_ERROR',

  GET_EVENTS_BY_MONTH_START: 'GET_EVENTS_BY_MONTH_START',
  GET_EVENTS_BY_MONTH_SUCCESS: 'GET_EVENTS_BY_MONTH_SUCCESS',
  GET_EVENTS_BY_MONTH_ERROR: 'GET_EVENTS_BY_MONTH_ERROR',
};

export default EventsTypes;
