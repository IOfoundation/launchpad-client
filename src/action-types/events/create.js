const EventsCreateTypes = {
  CREATE_EVENT_START: 'CREATE_EVENT_START',
  CREATE_EVENT_SUCCESS: 'CREATE_EVENT_SUCCESS',
  CREATE_EVENT_ERROR: 'CREATE_EVENT_ERROR',
};

export default EventsCreateTypes;
