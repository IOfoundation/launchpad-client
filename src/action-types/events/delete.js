const EventDeleteTypes = {
  DELETE_EVENT_START: 'DELETE_EVENT_START',
  DELETE_EVENT_SUCCESS: 'DELETE_EVENT_SUCCESS',
  DELETE_EVENT_ERROR: 'DELETE_EVENT_ERROR',
};

export default EventDeleteTypes;
