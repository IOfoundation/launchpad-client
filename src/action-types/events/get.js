const GetEventTypes = {
  GET_EVENT_START: 'GET_EVENT_START',
  GET_EVENT_SUCCESS: 'GET_EVENT_SUCCESS',
  GET_EVENT_ERROR: 'GET_EVENT_ERROR',

  SET_LOADING: 'SET_LOADING',
};

export default GetEventTypes;
