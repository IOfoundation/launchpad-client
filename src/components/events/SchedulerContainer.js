import React, {PureComponent} from 'react';
import '@progress/kendo-ui';
import '@progress/kendo-ui/js/kendo.timezones';
import {PropTypes} from 'prop-types';
import {Scheduler} from '@progress/kendo-scheduler-react-wrapper';
import {withStyles} from '@material-ui/core/styles';
import kendo from '@progress/kendo-ui';

import {containerStyles, getDate, mobileDaysMap} from '@Utils';

const styles = theme => ({
  container: {
    padding: 12,
    ...containerStyles(theme),
  },
  paper: {
    backgroundColor: theme.palette.background.paper,
    borderRadius: '5px',
    boxShadow: theme.shadows[5],
    left: '50%',
    maxHeight: '70vh',
    outline: 'none',
    overflow: 'auto',
    padding: '32px 24px',
    position: 'absolute',
    top: '45%',
    transform: 'translate(-50%, -50%)',
    width: theme.spacing.unit * 70,
    [theme.breakpoints.down('xs')]: {
      width: '100%',
      boxSizing: 'border-box',
    },
  },
});

class SchedulerContainer extends PureComponent {
  state = {
    views: ['day', 'week', {type: 'month', selected: true}],
    events: [],
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    if (
      nextProps.events.data &&
      nextProps.events.data.length !== prevState.events.length
    ) {
      const mappedEvents = nextProps.events.data.map(event => {
        const start = getDate(event.starting_at);
        const end = getDate(event.ending_at);
        const mapped = {
          ID: event.id,
          title: event.title,
          start: new Date(start.pacificTime),
          end: new Date(end.pacificTime),
          description: event.body,
          recurrenceId: event.organization_id,
        };

        if (event.is_all_day) {
          mapped.isAllDay = event.is_all_day;
        }

        return mapped;
      });

      return {events: mappedEvents};
    }

    return null;
  }

  componentDidMount() {
    this.props.actions.getAllEventsByMonth(this._date.monthNumeric);
    this._schedulerRef.widgetInstance.element[0].addEventListener(
      'click',
      this._schedulerClicked
    );
  }

  componentDidUpdate(prevProps) {
    const {data: nextEvents, loading: nextLoading} = this.props.events;
    const {data: currentEvents, loading: prevLoading} = prevProps.events;

    if (
      this._thereIsNoEventsAndIsNotLoading(
        nextEvents,
        currentEvents,
        nextLoading,
        prevLoading
      )
    ) {
      this.setLoadingStatusOnScheduler(false);
    }
  }

  componentWillUnmount() {
    this._schedulerRef.widgetInstance.element[0].removeEventListener(
      'click',
      this._schedulerClicked
    );
  }

  _schedulerRef = null;
  _schedulerClicked = event => {
    const className = event.target.className;

    if (className === 'k-event-template') {
      this.openModal(event.target.textContent);
    }
  };
  _thereIsNoEventsAndIsNotLoading = (
    nextEvents,
    currentEvents,
    nextLoading,
    prevLoading
  ) => {
    return (
      nextEvents.length === currentEvents.length &&
      nextEvents.length === 0 &&
      nextLoading !== prevLoading &&
      !nextLoading
    );
  };
  _changeName = true;
  _date = getDate();
  _month = this._date.monthNumeric;

  openModal = text => {
    const modalInformation = this.props.events.data.find(
      _event => _event.title === text
    );

    this.props.handlerModalVisibility(modalInformation);
  };

  updateCalendar = event => {
    const pickedDate = getDate(event.sender._model.selectedDate);
    const monthNumeric = pickedDate.date.toLocaleString('en-US', {
      month: 'numeric',
    });

    if (event.sender._selectedViewName === 'week' && this._changeName) {
      const weekDate = this._schedulerRef.widgetInstance.element[0].querySelectorAll(
        '.k-scheduler-header .k-nav-day'
      );

      weekDate.forEach(date => {
        const parse = date.textContent.split(' ');

        date.textContent = `${mobileDaysMap[parse[0]]} ${parse[1]}`;
      });
      this._changeName = false;
    } else if (event.sender._selectedViewName !== 'week') {
      this._changeName = true;
    }

    if (monthNumeric !== this._month) {
      this.setLoadingStatusOnScheduler(true);
      this._month = monthNumeric;
      this.props.actions.getAllEventsByMonth(monthNumeric);
    }
  };

  handlerDataBinding = event => {
    if (event.action === 'rebind' && event.items.length > 0) {
      this.setLoadingStatusOnScheduler(false);
    } else {
      this.setLoadingStatusOnScheduler(true);
    }
  };

  setLoadingStatusOnScheduler = value => {
    const scheduler = this._schedulerRef.widgetInstance.element.getKendoScheduler();
    kendo.ui.progress(scheduler.element.find('.k-scheduler-content'), value);
  };

  render() {
    const {classes, breakpoint} = this.props;
    const {startTime, views, events} = this.state;

    return (
      <div className={classes.container}>
        <Scheduler
          height={660}
          views={views}
          dataSource={events}
          date={this._date.date}
          startTime={startTime}
          mobile={breakpoint === 'xs'}
          editable={false}
          ref={_ref => {
            this._schedulerRef = _ref;
          }}
          dataBound={event => this.updateCalendar(event)}
          dataBinding={event => this.handlerDataBinding(event)}
        />
      </div>
    );
  }
}

SchedulerContainer.propTypes = {
  actions: PropTypes.shape({
    getAllEventsByMonth: PropTypes.func,
  }),
  breakpoint: PropTypes.string,
  classes: PropTypes.shape({
    container: PropTypes.string,
    paper: PropTypes.string,
  }),
  events: PropTypes.shape({
    data: PropTypes.arrayOf(PropTypes.shape({})),
    errors: PropTypes.shape({}),
    loading: PropTypes.bool,
  }),
  handlerModalVisibility: PropTypes.func,
};

export default withStyles(styles)(SchedulerContainer);
