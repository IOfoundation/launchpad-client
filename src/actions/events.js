import {EventsTypes as types} from '../action-types';
import {httpRequest, getDate} from '../utils';

const getAllEventsRequestStart = () => {
  return {
    type: types.GET_ALL_EVENTS_REQUEST_START,
  };
};

const getAllEventsRequestSuccess = data => {
  return {
    type: types.GET_ALL_EVENTS_REQUEST_SUCCESS,
    data,
  };
};

const getAllEventsRequestError = error => {
  return {
    type: types.GET_ALL_EVENTS_REQUEST_ERROR,
    error,
  };
};

const getEventsByMonthStart = () => {
  return {
    type: types.GET_EVENTS_BY_MONTH_START,
  };
};

const getEventsByMonthSuccess = featuredEvents => {
  return {
    type: types.GET_EVENTS_BY_MONTH_SUCCESS,
    featuredEvents,
  };
};

const getEventsByMonthError = error => {
  return {
    type: types.GET_EVENTS_BY_MONTH_ERROR,
    error,
  };
};

export const getAllEvents = () => {
  return async dispatch => {
    try {
      dispatch(getAllEventsRequestStart());
      const httpResponse = await httpRequest.get('/api/events');
      dispatch(getAllEventsRequestSuccess(httpResponse.data));
    } catch (error) {
      dispatch(getAllEventsRequestError(error));
    }
  };
};

const getAllEventsByMonthStart = () => {
  return {
    type: types.GET_ALL_EVENTS_BY_MONTH_START,
  };
};

const getAllEventsByMonthSuccess = eventsByMonth => {
  return {
    type: types.GET_ALL_EVENTS_BY_MONTH_SUCCESS,
    eventsByMonth,
  };
};

const getAllEventsByMonthError = error => {
  return {
    type: types.GET_ALL_EVENTS_BY_MONTH_ERROR,
    error,
  };
};

export const getAllEventsByMonth = month => {
  return async dispatch => {
    try {
      dispatch(getAllEventsByMonthStart());
      const httpResponse = await httpRequest.get(`/api/events?month=${month}`);
      const totalCount = parseInt(httpResponse.headers['x-total-count'], 10);

      if (totalCount > 200) {
        const totalHttpResponse = await httpRequest.get(
          `/api/events?month=${month}&perPage=${totalCount}`
        );
        dispatch(getAllEventsByMonthSuccess(totalHttpResponse.data));
      } else {
        dispatch(getAllEventsByMonthSuccess(httpResponse.data));
      }
    } catch (error) {
      dispatch(getAllEventsByMonthError(error));
    }
  };
};

export const getAllEventsById = id => {
  return async dispatch => {
    try {
      dispatch(getAllEventsRequestStart());
      const httpResponse = await httpRequest.get(
        `/api/events?organization_id=${id}`
      );
      dispatch(getAllEventsRequestSuccess(httpResponse.data));
    } catch (error) {
      dispatch(getAllEventsRequestError(error));
    }
  };
};

export const getEventsByMonth = (months, filterBy = 'default') => {
  return async dispatch => {
    try {
      dispatch(getEventsByMonthStart());

      const responses = await Promise.all(
        months.map(async month => {
          const url = {
            featured: `/api/events?month=${month}&featured=true`,
            default: `/api/events?month=${month}`,
          };

          const httpResponse = await httpRequest.get(url[filterBy]);

          const groupedData = httpResponse.data
            .slice(0, 3)
            .reduce((acc, data) => {
              data.starting_at = getDate(data.starting_at);
              data.posted_at = getDate(data.posted_at);
              acc.title = data.starting_at.monthLarge;
              acc.key = data.starting_at.month;

              if (acc[data.starting_at.month]) {
                acc[data.starting_at.month].push(data);
              } else {
                acc[data.starting_at.month] = [data];
              }

              return acc;
            }, {});

          return groupedData;
        })
      );

      dispatch(getEventsByMonthSuccess(responses));
    } catch (error) {
      dispatch(getEventsByMonthError(error));
    }
  };
};
