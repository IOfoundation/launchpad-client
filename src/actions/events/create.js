import {EventsCreateTypes as types} from '../../action-types';
import {httpRequest} from '@Utils';

const createEventStart = () => {
  return {
    type: types.CREATE_EVENT_START,
  };
};

const createEventSuccess = data => {
  return {
    type: types.CREATE_EVENT_SUCCESS,
    data,
  };
};

const createEventError = errors => {
  return {
    type: types.CREATE_EVENT_ERROR,
    errors,
  };
};

export const create = ({
  Authorization,
  organizationId: organization_id,
  event,
  mode,
}) => {
  return async dispatch => {
    try {
      dispatch(createEventStart());
      const config = {
        headers: {Authorization},
      };
      let httpResponse;

      if (mode === 'new') {
        httpResponse = await httpRequest.post(
          '/api/events',
          {...event, organization_id},
          config
        );
      } else if (mode === 'edit') {
        httpResponse = await httpRequest.put(
          '/api/events',
          {...event, organization_id},
          config
        );
      }

      dispatch(createEventSuccess(httpResponse.data));
    } catch (errors) {
      if (errors.data && errors.data.errors) {
        dispatch(createEventError(errors.data.errors));
      } else {
        dispatch(createEventError(errors.data));
      }
    }
  };
};
