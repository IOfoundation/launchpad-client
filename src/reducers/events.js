import {EventsTypes as types} from '../action-types';

const initialState = {
  data: [],
  eventsByMonth: [],
  featuredEvents: [],
};

export default function(state = initialState, action) {
  switch (action.type) {
    case types.GET_ALL_EVENTS_REQUEST_START: {
      return {
        ...state,
      };
    }

    case types.GET_ALL_EVENTS_REQUEST_SUCCESS: {
      const {data} = action;
      return {
        ...state,
        data,
      };
    }

    case types.GET_ALL_EVENTS_REQUEST_ERROR: {
      const {error} = action;
      return {
        ...state,
        error,
      };
    }

    case types.GET_EVENTS_BY_MONTH_START: {
      return {
        ...state,
      };
    }

    case types.GET_EVENTS_BY_MONTH_SUCCESS: {
      const {featuredEvents} = action;
      return {
        ...state,
        featuredEvents,
      };
    }

    case types.GET_EVENTS_BY_MONTH_ERROR: {
      const {error} = action;
      return {
        ...state,
        error,
      };
    }

    case types.GET_ALL_EVENTS_BY_MONTH_START: {
      return {
        ...state,
      };
    }

    case types.GET_ALL_EVENTS_BY_MONTH_SUCCESS: {
      const {eventsByMonth} = action;
      return {
        ...state,
        eventsByMonth,
      };
    }

    case types.GET_ALL_EVENTS_BY_MONTH_ERROR: {
      const {error} = action;
      return {
        ...state,
        error,
      };
    }

    default:
      return state;
  }
}
