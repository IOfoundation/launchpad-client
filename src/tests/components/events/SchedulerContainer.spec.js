/* eslint-env jest */
import React from 'react';
import {mount} from 'enzyme';
import thunk from 'redux-thunk';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';

import {EVENTS} from './mock/events';
import SchedulerContainer from '../../../components/events/SchedulerContainer';

jest.useFakeTimers();
/*eslint-disable */
Date.now = jest.fn(() => new Date(Date.UTC(2019, 0, 1, 0, 0, 0)));
Date = class extends Date {
  constructor(date) {
    if (date) {
      return super(date);
    }
    return new Date('05-08-89');
  }
};
/*eslint-enable */

const initialState = {
  events: {
    data: [...EVENTS],
    featuredEvents: [],
  },
};
const mockStore = configureStore([thunk]);
const props = {};
let store;
let wrapper;

describe('<SchedulerContainer />', () => {
  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <SchedulerContainer {...props} />
      </Provider>
    );
  });

  it('renders snapshot of SchedulerContainer', () => {
    global.Date = jest.fn().mockReturnValue('9/5/1989');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should get all the events', () => {
    expect(store.getActions()).toEqual([
      {type: 'GET_ALL_EVENTS_BY_MONTH_START'},
    ]);
  });
});
