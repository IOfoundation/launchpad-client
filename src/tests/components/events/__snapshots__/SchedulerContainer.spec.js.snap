// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SchedulerContainer /> renders snapshot of SchedulerContainer 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <WithStyles(Connect(SchedulerContainer))>
    <Connect(SchedulerContainer)
      classes={
        Object {
          "container": "Connect-SchedulerContainer--container-1",
          "paper": "Connect-SchedulerContainer--paper-2",
        }
      }
    >
      <SchedulerContainer
        actions={
          Object {
            "getAllEvents": [Function],
            "getAllEventsById": [Function],
            "getAllEventsByMonth": [Function],
            "getEventsByMonth": [Function],
          }
        }
        classes={
          Object {
            "container": "Connect-SchedulerContainer--container-1",
            "paper": "Connect-SchedulerContainer--paper-2",
          }
        }
      >
        <div
          className="Connect-SchedulerContainer--container-1"
        >
          <WithStyles(Modal)
            onClose={[Function]}
            open={false}
          >
            <Modal
              BackdropComponent={[Function]}
              classes={
                Object {
                  "hidden": "MuiModal-hidden-4",
                  "root": "MuiModal-root-3",
                }
              }
              closeAfterTransition={false}
              disableAutoFocus={false}
              disableBackdropClick={false}
              disableEnforceFocus={false}
              disableEscapeKeyDown={false}
              disablePortal={false}
              disableRestoreFocus={false}
              hideBackdrop={false}
              keepMounted={false}
              manager={
                ModalManager {
                  "data": Array [],
                  "handleContainerOverflow": true,
                  "hideSiblingNodes": true,
                  "modals": Array [],
                }
              }
              onClose={[Function]}
              open={false}
            />
          </WithStyles(Modal)>
          <Scheduler
            dataBound={[Function]}
            dataSource={Array []}
            date={1989-05-08T05:00:00.000Z}
            editable={false}
            height={660}
            mobile={false}
            views={
              Array [
                "day",
                "week",
                Object {
                  "selected": true,
                  "type": "month",
                },
              ]
            }
          >
            <div />
          </Scheduler>
        </div>
      </SchedulerContainer>
    </Connect(SchedulerContainer)>
  </WithStyles(Connect(SchedulerContainer))>
</Provider>
`;
